lib/beaker/host.rb

```diff
@@ -294,9 +294,25 @@ def is_x86_64?
 
     def connection
       # create new connection object if necessary
+      # Pull the connection information from the host configuration
+      ssh_opts = self['ssh']
+      ssh_opts = {} unless ssh_opts.is_a?(Hash)
+
+      if self.options['ssh']
+        if ssh_opts[:config]
+          # Load the configuration from the SSH config file
+          ssh_opts = self.options['ssh'].merge(Net::SSH::Config.load(ssh_opts[:config], self['ip']))
+        else
+          ssh_opts = self.options['ssh']
+        end
+      end
+
+      # Remove any options that don't apply to the Net::SSH library
+      ssh_opts = Net::SSH::Config.translate(ssh_opts)
```

Comment:
based on https://www.rubydoc.info/github/net-ssh/net-ssh/Net%2FSSH%2FConfig.translate I'm worried that if the keys are already symbols that it will remove them?

see that the code has a bunch of presets that have been removed: https://github.com/puppetlabs/beaker/blob/master/lib/beaker/options/presets.rb#L198-L206

In any case, if this parsing is necessary, would it not be only when we just loaded the new config via the file on line 304? If it is required in that case I would put it up within the `if` block only.

Reply:
I think I tried that and couldn't get it to work for some reason.

I guess the issue is that the underlying usage of Net::SSH isn't going to handle anything you might be passing to it anyway if it doesn't support it.

This means that all of the random things that are extra in the Hash may be useless (I think).

Reply:
I believe they are valid ssh options for example look at https://www.rubydoc.info/github/net-ssh/net-ssh/Net%2FSSH%2FConfig
:user_known_hosts_file is valid but is not present in the filanl hash. it could be because of line 306 (other comment). I can try some IRB on test hashes to see what exactly this translate does.

Reply:
it does remove everything in IRB:
```
irb(main):003:0> require 'net/scp'
=> true
irb(main):004:0> hash={:config=>false, :verify_host_key=>false, :auth_methods=>["publickey"], :port=>22, :forward_agent=>true, :keys=>["/var/lib/jenkins/.ssh/id_rsa-acceptance"], :user_known_hosts_file=>"/var/lib/jenkins/.ssh/known_hosts", :keepalive=>true}
=> {:config=>false, :verify_host_key=>false, :auth_methods=>["publickey"], :port=>22, :forward_agent=>true, :keys=>["/var/lib/jenkins/.ssh/id_rsa-acceptance"], :user_known_hosts_file=>"/var/lib/jenkins/.ssh/known_hosts", :keepalive=>true}
irb(main):005:0> hash[:config]
=> false
irb(main):006:0> Net::SSH::Config.translate(hash)
=> {:auth_methods=>["none", "publickey", "password", "keyboard-interactive"], :port=>22}
```

Also looking at the source code it seems to try to translate Openssh configs to net ssh configs, and I don't think we need that here.

Reply:
irb(main):003:0> require 'net/scp'
=> true
irb(main):004:0> hash={:config=>false, :verify_host_key=>false, :auth_methods=>["publickey"], :port=>22, :forward_agent=>true, :keys=>["***"], :user_known_hosts_file=>"***/known_hosts", :keepalive=>true}
=> {:config=>false, :verify_host_key=>false, :auth_methods=>["publickey"], :port=>22, :forward_agent=>true, :keys=>["****"], :user_known_hosts_file=>"****/known_hosts", :keepalive=>true}
irb(main):005:0> hash[:config]
=> false
irb(main):006:0> Net::SSH::Config.translate(hash)
=> {:auth_methods=>["none", "publickey", "password", "keyboard-interactive"], :port=>22}

Reply:
Ah...the keys have to be strings :-/

https://net-ssh.github.io/ssh/v2/api/classes/Net/SSH/Config.html#M000302

Reply:
yes look at the source code, not only strings but the terms are slightly different eg. `userknownhostsfile` becomes `:user_known_hosts_file`
`pubkeyauthentication` becomes an element of the :auth_methods hash
etc.